src/Prove.hs

```diff
@@ -0,0 +1,340 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Prove (queries) where
+
+import Control.Monad (when)
+import Data.ByteString (ByteString)
+import Data.ByteString.UTF8 (toString)
+import Data.List (intercalate, (\\), nub)
+import Data.List.NonEmpty as NonEmpty (toList)
+import Data.Map.Strict as Map (Map, lookup, fromList)
+import Data.Maybe
+
+import Data.SBV hiding (name)
+import Data.SBV.String ((.++), subStr)
+
+import RefinedAst
+import Syntax (Id, Interface(..), Decl(..))
+import Type (metaType, mkStorageBounds)
+
+-- *** Interface *** --
+
+
+{-|
+   For each Invariant claim build an SMT query consisting of:
+
+   - constants for the pre and post versions of all storage variables used by the transition system
+   - boolean predicates over the pre and post storage variables for each pass behaviour
+   - a boolean predicate for the invariant
+   - an assertion that the invariant does not hold if either of the following is true:
+      1. The constructor predicate holds
+      2. The invariant holds over the prestate and one of the method level predicates holds
+
+   If this query returns `unsat` then the invariant must hold over the transition system
+-}
+queries :: [Claim] -> [(Invariant, [Symbolic ()])]
```

Comment:
what do you think of making this a Map rather than a list of pairs? It might be easier to read. And we could use the functor instance for map to map over its values like you are doing on https://github.com/ethereum/act/pull/37/files#diff-f8f3412da88cd4806f23d59fe59ebc3bR111.
It would add some `fromList` and `toList` instances here and there though

Reply:
This needs an `Ord` instnace for `Invariant`, and by extension `Exp Bool`, which the compiler is not able to derive by itself, I'm not sure if it's worth the boilerplate?

What do you think?

Reply:
Right, ok, let's leave it as is